onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -391,6 +391,112 @@ TEST(CApiTest, create_session_without_session_option) {
+  Ort::Session session(*ort_env, MODEL_URI, session_options);
+
+  Ort::MemoryInfo info_cpu = Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemTypeDefault);
+
+  const std::array<int64_t, 2> x_shape = {3, 2};
+  std::array<float, 3 * 2> x_values  = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
+  Ort::Value bound_x = Ort::Value::CreateTensor(info_cpu, x_values.data(), x_values.size(),
+    x_shape.data(), x_shape.size());
+
+  const std::array<float, 3 * 2> expected_y = {1.0f, 4.0f, 9.0f, 16.0f, 25.0f, 36.0f};
+  const std::array<int64_t, 2> y_shape = {3, 2};
+  std::array<float, 3 * 2> y_values;
+  Ort::Value bound_y = Ort::Value::CreateTensor(info_cpu, y_values.data(), y_values.size(),
+                                                y_shape.data(), y_shape.size());
+
+
+  Ort::IoBinding binding(session);
+  binding.BindInput("X", bound_x);
+  binding.BindOutput("Y", bound_y);
+
+  session.Run(Ort::RunOptions(), binding);
+  // Check the values against the bound raw memory
+  ASSERT_TRUE(std::equal(std::begin(y_values), std::end(y_values), std::begin(expected_y)));
+
+  // Now compare values via GetOutputValues
+  {
+    std::vector<Ort::Value> output_values = binding.GetOutputValues();
+    ASSERT_EQ(output_values.size(), 1U);
+    const Ort::Value& Y_value = output_values[0];
+    ASSERT_TRUE(Y_value.IsTensor());
+    Ort::TensorTypeAndShapeInfo type_info = Y_value.GetTensorTypeAndShapeInfo();
+    ASSERT_EQ(ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT, type_info.GetElementType());
+    auto count = type_info.GetElementCount();
+    ASSERT_EQ(expected_y.size(), count);
+    const float* values = Y_value.GetTensorData<float>();
+    ASSERT_TRUE(std::equal(values, values + count, std::begin(expected_y)));
+  }
+
+  {
+    std::vector<std::string> output_names = binding.GetOutputNames();
+    ASSERT_EQ(1U, output_names.size());
+    ASSERT_EQ(output_names[0].compare("Y"), 0);
+  }
+
+  // Now replace binding of Y with an on device binding instead of pre-allocated memory.
+  // This is when we can not allocate an OrtValue due to unknown dimensions
+  {
+    Ort::MemoryInfo info_cpu_dev("Cpu", OrtAllocatorType::OrtArenaAllocator, 0, OrtMemTypeDefault);
+    binding.BindOutput("Y", info_cpu_dev);
+    session.Run(Ort::RunOptions(), binding);
+  }
+
+  // Check the output value allocated based on the device binding.
+  {
+    std::vector<Ort::Value> output_values = binding.GetOutputValues();
+    ASSERT_EQ(output_values.size(), 1U);
+    const Ort::Value& Y_value = output_values[0];
+    ASSERT_TRUE(Y_value.IsTensor());
+    Ort::TensorTypeAndShapeInfo type_info = Y_value.GetTensorTypeAndShapeInfo();
+    ASSERT_EQ(ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT, type_info.GetElementType());
+    auto count = type_info.GetElementCount();
+    ASSERT_EQ(expected_y.size(), count);
+    const float* values = Y_value.GetTensorData<float>();
+    ASSERT_TRUE(std::equal(values, values + count, std::begin(expected_y)));
+  }
+
+  binding.ClearBoundInputs();
```

Comment:
What will this do? Will it release the input/output value?

Reply:
This will clear the binding. Yes, it will release input and output value. However, the values released may or may not own the buffer. If you bind to a pre-allocated memory, then you are responsible for de-allocating it.

---
In reply to: [463792543](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r463792543) [](ancestors = 463792543)

Reply:
I see. Then why not do this in the destructor of `IoBinding`, like`Ort::Value`?

Reply:
We can't do it as the allocation may come from a custom or 3rd party allocator. There are no memory leaks in modern C++.

---
In reply to: [464504880](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r464504880) [](ancestors = 464504880)

Reply:
I think my response may not have been clear enough.
ClearBoundInputs/Outputs() remove the binding but it has no effect the Value instance you pass in since Bind() calls make a copy of Value object.
Value destructor will take care of itslef and of the memory it owns if it owns it.

---
In reply to: [464556707](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r464556707) [](ancestors = 464556707,464504880)

Reply:
For you concern of dealing with memory leaks, DefaultAllocator and Allocator now have GetAllocation() method that return MemoryAllocation  instance. Upon destuction it would properly deallocate memory.

---
In reply to: [464612358](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r464612358) [](ancestors = 464612358,464556707,464504880)

Reply:
Sorry, I am still a little confused. For the user, should we explicitly call ClearBoundOutput/ClearBoundInputs since the destructor already destroyed the internal Values? 

For the Values passed from outside, I know outside will destroy it, for the case of `BindOutputToDevice`, the output tensor will be released when the IoBinding object is out of scope, right? Or we need to call ClearBoundOutput explicitly?

Reply:
ClearBoundInputs/Outputs are convenience methods. You do not have to use them. They only make sense if you want to remove some bindings or start anew. They deal with the internal structures.